HomeWork 6 and 6.1 Robot

andstartut opened the pull request.

andstartut committed:
HomeWork 6 and 6.1 Robot

GlebLitvinov commented on Lesson6.1/src/by/teachmeskills/robot/Run.java:
```diff
@@ -0,0 +1,38 @@
+            System.out.println("The Vityaz is the most expensive robot");
+        } else if (integral.getPrice() > horizont.getPrice() & integral.getPrice() > vityaz.getPrice()) {
+            System.out.println("The Integral is the most expensive robot");
+        }
```
1) please move this logic out of Run class to make your code look more likely to OOP style
2) what if you will have 4 robots? or 6? refactor your algorithm to work with an array of robots

GlebLitvinov reviewed (changes requested):
exceptions and robots core logic is good, but you need to refactor Run class

andstartut committed:
Rebuilt the Robot project.
1) Used the general RobotPart class for Head, Hand and Leg.
2) Moved all implementation from the Main class.
3) Used a factories classes to created new parts and robots.
4) Used a Enum classes and fields to fill up a name of parts, brands, prices and markups.

GlebLitvinov commented on Lesson6.1/src/by/teachmeskills/robot/parts/RobotPart.java:
```diff
@@ -0,0 +1,49 @@
+        return convertingConstantToName(part.name());
+    }
+
+    /**
```
yeah, you can create smth like RobotNameUtil and move this method there

GlebLitvinov commented on Lesson6.1/src/by/teachmeskills/robot/parts/RobotPart.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public void info() {
+        System.out.println(getPartName() + "'s " + getBrandName() + " is moved");
```
this is an interesting implementation, but what if different parts (legs/heads/...) will have their own behavior? Hand should wave, leg should step e.t.c.

GlebLitvinov commented on Lesson6.1/src/by/teachmeskills/robot/factory/RandomRobotPartFactory.java:
```diff
@@ -0,0 +1,33 @@
+
+            case 2 -> brand = TOSHIBA;
+        }
+        return brand;
```
it can be replaced to return BrandsAndMarkupIndex.values()[random.nextInt(ENUM_LENGTH)];

andstartut committed:
Les6 clean

GlebLitvinov commented on Lesson6.1/src/by/teachmeskills/robot/Robot.java:
```diff
@@ -0,0 +1,53 @@
+    private IRobotPart hand;
+    private IRobotPart leg;
+
+    public Robot(IRobotPart head, IRobotPart hand, IRobotPart leg) {
```
writing it like this would allow to create a robot using 3 hands. I don't think that this is a good practice

GlebLitvinov replied:
if you want to leave it as it is, you need to make robot fabric the only place that can create robots

andstartut committed:
les 6.1
Added the RobotNameUtil class with different info implementation.

GlebLitvinov commented on Lesson6.1/src/by/teachmeskills/robot/parts/RobotNameUtil.java:
```diff
@@ -0,0 +1,26 @@
+
+public class RobotNameUtil {
+
+    public final String convertingConstantToName(String constantName) {
```
you can make these methods static, and then you won't need to create object of RobotNameUtil in every robot, you will be able to call this method like RobotNameUtil. convertingConstantToName(name)

andstartut committed:
les 6.1
The robot fabric is the only place that can create robots.

andstartut committed:
les 6.1
RobotNameUtil class made static

andstartut committed:
Merge remote-tracking branch 'origin/Lesson6' into Lesson6

# Conflicts:
#	Lesson6.1/src/by/teachmeskills/robot/parts/RobotPart.java

andstartut committed:
les 6.1
RobotNameUtil class made static

andstartut committed:
les 6.1
Added CreateRobotPartException
